h2o-core/src/main/java/water/parser/ParseDataset.java

```diff
@@ -714,15 +714,18 @@ private FVecParseWriter makeDout(ParseSetup localSetup, int chunkOff, int nchunk
       final int chunkStartIdx = _fileChunkOffsets[_lo];
       Log.trace("Begin a map stage of a file parse with start index " + chunkStartIdx + ".");
 
+      DecryptionTool decryptionTool = DecryptionTool.get(_parseSetup._decrypt_tool);
       byte[] zips = vec.getFirstBytes();
       ZipUtil.Compression cpr = ZipUtil.guessCompressionMethod(zips);
-      if (localSetup._check_header == ParseSetup.HAS_HEADER) //check for header on local file
-        localSetup._check_header = localSetup.parser(_jobKey).fileHasHeader(ZipUtil.unzipBytes(zips,cpr, localSetup._chunk_size), localSetup);
+      if (localSetup._check_header == ParseSetup.HAS_HEADER) { //check for header on local file
+        byte[] bits = decryptionTool.decryptFirstBytes(ZipUtil.unzipBytes(zips, cpr, localSetup._chunk_size));
+        localSetup._check_header = localSetup.parser(_jobKey).fileHasHeader(bits, localSetup);
+      }
       // Parse the file
       try {
         switch( cpr ) {
         case NONE:
-          ParserInfo.ParseMethod pm = _parseSetup._parse_type.parseMethod(_keys.length,vec.nChunks());
```

Comment:
I think it would be better to make special case (or add if statement here) for encrypted-non compressed instead of passing extra boolean flag to parseMethod. Parser(parserInfo) has no knowledge of compression and no way to decide on its own what parse method to use . Also Orc parser overrides parseMethod (it misses the @override annotation and is only a performance hack so it does not fail) and should be changed as well (and @Override should be added in either case ;).

Reply:
That is a good point, thanks. I will fix it.